### PR TITLE
Added C++ API test

### PR DESCRIPTION
**Pull request by izdeby**
This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

--------------
In this PR: 
Add a test to check that C++ API behavior stays the same after all the changes. 
While working on it a bug related to `requires_grad` was found and logged in the master task.

--------------
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30983 Updated codegen to remove TensorOptions from generated code
* #30982 Prepare templates
* #30981 Updated DispatchKeyExtractor to expect TensorOptions
* **#30980 Added C++ API test**
* #30979 Add tracing support for optional Device and Layout

Differential Revision: [D18912681](https://our.internmc.facebook.com/intern/diff/D18912681)

### CODE REVIEW

File: aten/src/ATen/test/basic.cpp [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 #include <ATen/ATen.h>
 #include <ATen/core/Reduction.h>
+#include <torch/cuda.h>
```
Comment (gchanan) [inline aten/src/ATen/test/basic.cpp]: what do you need from this?

File: aten/src/ATen/test/basic.cpp [hunk 2]
```diff
@@ -345,3 +346,69 @@ TEST(BasicTest, BasicTestCUDA) {
...
+TEST(BasicTest, FactoryMethodsTest) {
+  // Test default values
+  at::Tensor tensor0 = at::empty({4});
+  ASSERT_EQ(tensor0.dtype(), at::kFloat);
+  ASSERT_EQ(tensor0.layout(), at::kStrided);
+  ASSERT_EQ(tensor0.device(), at::kCPU);
+  ASSERT_FALSE(tensor0.requires_grad());
+  ASSERT_FALSE(tensor0.is_pinned());
+
+  // Test setting grad to FALSE
```
Comment (gchanan) [inline aten/src/ATen/test/basic.cpp]: nit: Test setting requires_grad to false.

File: aten/src/ATen/test/basic.cpp [hunk 3]
```diff
@@ -345,3 +346,69 @@ TEST(BasicTest, BasicTestCUDA) {
...
+  tensor0 = at::empty({4}, at::TensorOptions().requires_grad(false));
+  ASSERT_EQ(tensor0.dtype(), at::kFloat);
+  ASSERT_EQ(tensor0.layout(), at::kStrided);
+  ASSERT_EQ(tensor0.device(), at::kCPU);
+  ASSERT_FALSE(tensor0.requires_grad());
+  ASSERT_FALSE(tensor0.is_pinned());
+
+  // Test setting grad to TRUE
```
Comment (gchanan) [inline aten/src/ATen/test/basic.cpp]: requires_grad.

File: aten/src/ATen/test/basic.cpp [hunk 4]
```diff
@@ -345,3 +346,69 @@ TEST(BasicTest, BasicTestCUDA) {
...
+    ASSERT_EQ(tensor1.dtype(), at::kHalf);
+    ASSERT_EQ(tensor1.layout(), at::kSparse);
+    ASSERT_TRUE(tensor1.device().is_cuda());
+    ASSERT_FALSE(tensor1.requires_grad());
+
+    // This will cause an exception. 
+    // Issue #30405
+    ASSERT_ANY_THROW(tensor1.is_pinned());
+  }
+}
```
Comment (gchanan) [inline aten/src/ATen/test/basic.cpp]: newline.

File: aten/src/ATen/test/basic.cpp [hunk 5]
```diff
@@ -345,3 +346,69 @@ TEST(BasicTest, BasicTestCUDA) {
...
+  ASSERT_FALSE(tensor0.requires_grad());
+  ASSERT_FALSE(tensor0.is_pinned());
+
+  // Test setting grad to TRUE
+  tensor0 = at::empty({4}, at::TensorOptions().requires_grad(true));
+  ASSERT_EQ(tensor0.dtype(), at::kFloat);
+  ASSERT_EQ(tensor0.layout(), at::kStrided);
+  ASSERT_EQ(tensor0.device(), at::kCPU);
+  // This is a bug. Requires_grad was set to TRUE but this is being ignored.
+  // Issue #30405
```
Comment (gchanan) [inline aten/src/ATen/test/basic.cpp]: include the entire github link.

File: aten/src/ATen/test/basic.cpp [hunk 6]
```diff
@@ -345,3 +346,69 @@ TEST(BasicTest, BasicTestCUDA) {
...
+
+    // Test set everything
+    tensor1 = at::empty({4}, at::TensorOptions().dtype(at::kHalf).device(at::kCUDA).layout(at::kSparse).requires_grad(true));
+    ASSERT_EQ(tensor1.dtype(), at::kHalf);
+    ASSERT_EQ(tensor1.layout(), at::kSparse);
+    ASSERT_TRUE(tensor1.device().is_cuda());
+    ASSERT_FALSE(tensor1.requires_grad());
+
+    // This will cause an exception. 
+    // Issue #30405
```
Comment (gchanan) [inline aten/src/ATen/test/basic.cpp]: include the entire github link.

File: aten/src/ATen/test/basic.cpp [hunk 7]
```diff
@@ -345,3 +346,69 @@ TEST(BasicTest, BasicTestCUDA) {
...
+    ASSERT_TRUE(tensor1.device().is_cuda());
+    ASSERT_FALSE(tensor1.requires_grad());
+    ASSERT_FALSE(tensor1.is_pinned());
+
+    // Test set everything
+    tensor1 = at::empty({4}, at::TensorOptions().dtype(at::kHalf).device(at::kCUDA).layout(at::kSparse).requires_grad(true));
+    ASSERT_EQ(tensor1.dtype(), at::kHalf);
+    ASSERT_EQ(tensor1.layout(), at::kSparse);
+    ASSERT_TRUE(tensor1.device().is_cuda());
+    ASSERT_FALSE(tensor1.requires_grad());
```
Comment (gchanan) [inline aten/src/ATen/test/basic.cpp]: relink the broken issue here.

File: aten/src/ATen/test/basic.cpp (hunk 1 above)
Comment (izdeby) [reply aten/src/ATen/test/basic.cpp]: for `torch::cuda::is_available()`

File: aten/src/ATen/test/basic.cpp [hunk 8]
```diff
@@ -345,3 +346,72 @@ TEST(BasicTest, BasicTestCUDA) {
...
+    ASSERT_FALSE(tensor1.is_pinned());
+
+    // Test set everything
+    tensor1 = at::empty({4}, at::TensorOptions().dtype(at::kHalf).device(at::kCUDA).layout(at::kSparse).requires_grad(true));
+    ASSERT_EQ(tensor1.dtype(), at::kHalf);
+    ASSERT_EQ(tensor1.layout(), at::kSparse);
+    ASSERT_TRUE(tensor1.device().is_cuda());
+
+    // This is a bug
+    // Issue https://github.com/pytorch/pytorch/issues/30405
```
Comment (ezyang) [inline aten/src/ATen/test/basic.cpp]: Oof what a nasty bug

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/test/basic.cpp (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/test/basic.cpp]: can you follow what the other tests in this directory do?  Namely, use `at::cuda::is_available()` (and don't include this file).

Comment (izdeby) [reply aten/src/ATen/test/basic.cpp]: I cant use `at::cuda::is_available()` here as i will get an error: `error: ‘is_available’ is not a member of ‘at::cuda’`. Other files that use `at::cuda::is_available()` are including `<ATen/cuda/CUDAContext.h>` that enables this. I picked `<torch/cuda.h>` as its much easier to add than `<ATen/cuda/CUDAContext.h>` which will require additional work.
I can leave it as it is or move the test to another file, possibly CUDA test where its already enabled. What do you prefer?

Comment (gchanan) [reply aten/src/ATen/test/basic.cpp]: why does that involve additional work?  You should do what the other tests do here or if your way is better, change the other tests to do things the new way.  Not have two ways of doing the same thing.

**Pull request closed without merging**